### PR TITLE
Issue 2697

### PR DESCRIPTION
**Pull request by rppawlo**
Kokkosify Scatter Dirichlet BCs for Blocked Tpetra

Also contains a commit to fix some cuda warnings in the ZeroSensitivities functions.

Passes on OpenMP and CUDA
Clean on valgrind

### CODE REVIEW

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp [hunk 1]
```diff
@@ -205,109 +247,77 @@ preEvaluate(typename TRAITS::PreEvalData d)
...
+     const auto& tpetraDirichletCounter = *((rcp_dynamic_cast<Thyra::TpetraVector<RealType,LO,GO,NodeT>>(dirichletCounter_->getNonconstVectorBlock(productVectorBlockIndex_[fieldIndex]),true))->getTpetraVector());
+     const auto& kokkosDirichletCounter = tpetraDirichletCounter.template getLocalView<PHX::mem_space>();
+
+     // Class data fields for lambda capture
+     const auto& fieldOffsets = fieldOffsets_[fieldIndex];
+     const auto& basisIndices = basisIndexForMDFieldOffsets_[fieldIndex];
+     const auto& worksetLIDs = worksetLIDs_;
+     const auto& fieldValues = scatterFields_[fieldIndex];
+     const auto& applyBC = applyBC_[fieldIndex].get_static_view();
+     const bool checkApplyBC = checkApplyBC_;
```
Comment (jmgate) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: What's the motivation for this block?  Can we not access member data within a `Kokkos::parallel_for`?

Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: not on cuda if using lambda capture

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp [hunk 2]
```diff
@@ -411,170 +480,176 @@ preEvaluate(typename TRAITS::PreEvalData d)
...
+        // HACK to enforce views in the CrsGraph to be
+        // Unmanaged. Passing in the MemoryTrait<Unmanaged> doesn't
+        // work as the CrsGraph in the CrsMatrix ignores the
+        // MemoryTrait. Need to use the runtime constructor by passing
+        // in points to ensure Unmanaged.  See:
+        // https://github.com/kokkos/kokkos/issues/1581
+
+        // These two lines are the original code we can revert to when #1581 is fixed.
+        // const auto crsMatrix = rcp_dynamic_cast<Tpetra::CrsMatrix<double,LO,GO,NodeT>>(thyraTpetraOperator->getTpetraOperator(),true);
+        // new (&hostJacTpetraBlocks(row,col)) KokkosSparse::CrsMatrix<double,LO,PHX::Device,Kokkos::MemoryTraits<Kokkos::Unmanaged>> (crsMatrix->getLocalMatrix());
```
Comment (jmgate) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: Do we want to create an issue to remind us to switch this at some point?

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp (hunk 1 above)
Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: Oh - should also mention - when we can jump Trilinos to c++17 this should be fixed.

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp (hunk 2 above)
Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: yes

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp [hunk 3]
```diff
@@ -368,26 +378,85 @@ ScatterDirichletResidual_BlockedTpetra(const Teuchos::RCP<const BlockedDOFManage
...
+  // evaluate call is large enough to hold all derivatives for each
+  // sub block load
+  for (int blk=0;blk<numBlocks;blk++) {
+    const int blockDerivativeSize = hostBlockOffsets(blk+1) - hostBlockOffsets(blk);
+    TEUCHOS_TEST_FOR_EXCEPTION(blockDerivativeSize > 256, std::runtime_error,
+                               "ERROR: the derivative dimension for sub block "
+                               << blk << "with a value of " << blockDerivativeSize
+                               << "is larger than the size allocated for cLIDs and vals "
+                               << "in the evaluate call! You must manually increase the "
+                               << "size and recompile!");
```
[the remark this answers is not in the document]
Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: Value comes from original non-blocked version of code. Used to allocate temporary in kokos lambda. Looks like constexpr works in cuda 8. Let me experiment with that.

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp (hunk 1 above)
Comment (jmgate) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: Lambda capture works differently in C++17?  I'll set myself a reminder for a week before I retire&hellip;

Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_BlockedTpetra_impl.hpp]: The kokkos team successfully got support for *this capture into the 17 standard:

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0018r3.html

**Comment by mhoemmen** [comment]
Oops, I merged the wrong one!  It did pass PR testing though.

**Pull request merged**